.github/workflows/xp.yml

```diff
@@ -0,0 +1,359 @@
+          --health-retries 5
+        ports:
+          - 5432:5432
+    steps:
+      - name: Check out code
+        uses: actions/checkout@v3
+        with:
+          fetch-depth: 0
+
+      - name: Setup Test Dependencies
+        uses: ./.github/actions/setup-tests
+      
+      - name: Run Treatment Service test
+        env:
+          DATABASE_HOST: localhost
+          DATABASE_NAME: xp
+          DATABASE_USER: xp
+          DATABASE_PASSWORD: xp
+        run: make test-treatment-service
+
+  e2e-tests:
+    runs-on: ubuntu-latest
+    env:
+      GOPATH: ${{ github.workspace }}/api/.go
+    services:
+      postgres:
+        image: postgres:13-alpine
+        env:
+          POSTGRES_DB: xp
+          POSTGRES_USER: xp
+          POSTGRES_PASSWORD: xp
+        options: >-
+          --health-cmd pg_isready
+          --health-interval 10s
+          --health-timeout 5s
+          --health-retries 5
+        ports:
+          - 5432:5432
+    steps:
+      - name: Check out code
+        uses: actions/checkout@v3
+        with:
+          fetch-depth: 0
+
+      - name: Setup Test Dependencies
+        uses: ./.github/actions/setup-tests
+      
+      - name: Setup Python
+        id: setup-python
+        uses: actions/setup-python@v2
+        with:
+          python-version: "3.7"
+          architecture: x64
+
+      - name: Install dependencies
+        run: make install-python-ci-dependencies
+      
+      - name: Build binaries
+        run: make build
+
+      - name: Run E2E tests
+        env:
+          DATABASE_HOST: localhost
+          DATABASE_NAME: xp
+          DATABASE_USER: xp
+          DATABASE_PASSWORD: xp
+        run: make e2e-ci
+
+  build-ui:
+    runs-on: ubuntu-latest
+    defaults:
+      run:
+        working-directory: ui
+    steps:
+      - name: Checkout to the target branch
+        uses: actions/checkout@v3
+
+      - uses: actions/setup-node@v3
+        with:
+          node-version: 14
+
+      - name: Get yarn cache directory path
+        id: yarn-cache-dir-path
+        run: echo "::set-output name=dir::$(yarn cache dir)"
+
+      - name: Cache YARN
+        uses: actions/cache@v3
+        id: yarn-cache # use this to check for `cache-hit` (`steps.yarn-cache.outputs.cache-hit != 'true'`)
+        with:
+          path: ${{ steps.yarn-cache-dir-path.outputs.dir }}
+          key: ${{ runner.os }}-yarn-${{ hashFiles('**/yarn.lock') }}
+          restore-keys: |
+            ${{ runner.os }}-yarn-
+
+      - name: Install
+        run: yarn install --network-concurrency 1
+
+      - name: Lint code
+        run: yarn lint
+
+      - name: Build UI
+        env:
+          NODE_OPTIONS: "--max_old_space_size=4096"
+        run: yarn build
+
+      - name: Publish Artifact
+        uses: actions/upload-artifact@v3
+        with:
+          name: xp-ui-dist
+          path: ui/build/
+          retention-days: ${{ env.ARTIFACT_RETENTION_DAYS }}
+
+  build-management-service:
```

Comment:
Can we combine this with the publish step ? In turing, the build / publish separation makes sense because the built docker images are used in the e2e tests. We don't need to build the docker images ahead of time for XP, right? The built API binaries from the `e2e-tests` job can be saved as artifacts for reuse. Then, during the publish phase, we can build the UI and subsequently use the API binaries in the docker image for the management service. Similar steps for the treatment service minus the UI.